### DISCUSSION TITLE
useNodeState and useEdgesState hooks imported from reactflow cannot be used in production?

### DISCUSSION THREAD

**Discussion by nithish523** [xyflow/xyflow#2931] [Q&A]
I have noticed recently in the documentation that the `useNodeState` and `useEdgesState` hooks cannot be used in production. Also saw the warning that these hooks are used just to reduce the boilerplate code in providing the examples.
<img width="1442" alt="Screenshot 2023-03-20 at 12 13 58 AM" src="https://user-images.githubusercontent.com/111063820/226199910-6653538c-67b8-4b4a-b726-60c91fb72ffc.png">

I have also gone through the core file of this hook and I saw both the `useNodeState` and `useEdgesState` hooks use '`applyNodeChanges`' and '`applyEdgeChanges`' functions respectively to apply changes and also they both have the React state to store `nodes` and `edges` (not sure why this is not prod safe)
https://github.com/wbkd/react-flow/blob/674127a3eb6d2a70ca5894dffa7c5bad9d9769d5/packages/core/src/hooks/useNodesEdgesState.ts

The other way of setting nodes and edges I see in the documentation is by using `reactflow` instance
https://reactflow.dev/docs/api/react-flow-instance/#setnodes

Should I use `reactflow` instance to set nodes and edges for production mode?
Well, I don't see any reasons mentioned in the documentation why `useNodeState` and `useEdgesState` cannot be used for production. May i know the reason? Can you also suggest what would be the best approach?

**Comment by bcakmakoglu** [comment]
They *can* be used for production but it is recommended to use a proper state management library instead.
You could use [Zustand](https://github.com/pmndrs/zustand) for example.
